common/lockmanager_test.go

```diff
@@ -0,0 +1,318 @@
+    var firstThreadWaitGroup sync.WaitGroup;
+    var secondThreadWaitGroup sync.WaitGroup;
+    var bothThreadsWaitGroup sync.WaitGroup;
+
+    firstThreadWaitGroup.Add(1);
+    secondThreadWaitGroup.Add(1);
+    bothThreadsWaitGroup.Add(1);
+
+    // First thread aquires the lock a first time.
+    go func() {
+        defer firstThreadWaitGroup.Done();
+        Lock(key1);
+    }()
+
+    // Wait for the first thread to lock key1. 
+    firstThreadWaitGroup.Wait();
+
+
+    // Second thread aquires the lock a second time.
+    go func() {
+        defer bothThreadsWaitGroup.Done();
+        secondThreadWaitGroup.Done();
+        Lock(key1);
+    }()
+
+    // Wait for the second thread to try to lock key1
+    // while the lock is being used by the first thread.
+    secondThreadWaitGroup.Wait();
+    
+    // Small sleep to ensure the second thread gets to locking key1.
+    time.Sleep(10 * time.Millisecond);
+
+    // Unlock the first threads lock.
+    err := Unlock(key1);
+    if (err != nil) {
+        test.Fatalf("Failed to unlock.");
+    }
+    
+    // Wait for the second thread to aquire the lock.
+    bothThreadsWaitGroup.Wait();
+
+    // Unlock the second threads lock.
+    err = Unlock(key1);
+    if (err != nil) {
+        test.Fatalf("Failed to unlock.");
+    }
+}
+
+func TestLockUnlockSameKeyConcurrently(test *testing.T) {
+    key1 := "testkey1";
+    var threadWaitGroup sync.WaitGroup;
+    var verifyWaitGroup sync.WaitGroup;
+
+    // Lock key1 for the first time.
+    Lock(key1);
+
+    threadWaitGroup.Add(1);
+    verifyWaitGroup.Add(1);
+
+    // This thread should wait to lock/unlock key1 a second time until it gets
+    // unlocked for the first time.
+    go func() {
+        defer threadWaitGroup.Done();
+        verifyWaitGroup.Done();
+        Lock(key1);
+        defer func() {
+            err := Unlock(key1);
+            if (err != nil) {
+                test.Fatalf("Failed to unlock.");
+            }
+        }()
+    }()
+
+    // Make sure the thread is trying to Lock with the same key
+    // before the main thread unlocks it.
+    verifyWaitGroup.Wait();
+
+    // Small sleep to ensure the thread gets to locking key1.
+    time.Sleep(10 * time.Millisecond);
+
+    // Unlock key1 for the first time.
+    err := Unlock(key1);
+    if (err != nil) {
+        test.Fatalf("Failed to unlock.");
+    }
+
+    // Wait for the thread to finish locking and unlocking key1.
+    threadWaitGroup.Wait();
+}
+
+func TestStaleLockRetention(test *testing.T) {
+    key1 := "testkey1";
+    staleDuration := time.Duration(config.STALELOCK_DURATION_SECS.Get()) * time.Second;
+    var threadWaitGroup sync.WaitGroup;
+    var timestampWaitGroup sync.WaitGroup;
+
+    Lock(key1);
+    Unlock(key1);
+
+    // Load the lock data.
+    val, _ := lockMap.Load(key1); 
+    // Set the timestamp for a lock to be considered stale.
+    val.(*lockData).timestamp = time.Now().Add(-1 * (staleDuration + (1 * time.Second)));
```

Comment:
If we are making the lock stale, then go ahead and make it really stale.
Just double the stale time.